clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperation/SerializeApplicationAtomicOperation.groovy

```diff
@@ -0,0 +1,645 @@
+
+
+  /* curl -X POST -H "Content-Type: application/json" -d '[ { "serializeApplication": { "applicationName": "codelab", "credentials": "my-google-account" }} ]' localhost:7002/gce/ops */
+  @Override
+  Void operate(List priorOutputs) {
+
+    //TODO: static ip addresses
+    def resourceMap = [:]
+    initializeResourceMap(resourceMap)
+
+    googleClusterProvider.getClusters(applicationName, accountName).each { GoogleCluster.View cluster ->
+      cluster.serverGroups.each { GoogleServerGroup.View serverGroup ->
+        addServerGroupToResourceMap(serverGroup, resourceMap)
+      }
+    }
+
+    googleLoadBalancerProvider.getApplicationLoadBalancers(applicationName).each { GoogleLoadBalancer.View loadBalancer ->
+      addLoadBalancerToResourceMap(loadBalancer, resourceMap)
+    }
+
+    //TODO: get only the appropriate security groups rather than all
+    googleSecurityGroupProvider.getAll(true).each {GoogleSecurityGroup securityGroup ->
+      addSecurityGroupToResourceMap(securityGroup,resourceMap)
+    }
+
+    cleanUpResourceMap(resourceMap)
+
+    def json = new JsonBuilder()
+    def applicationMap = [:]
+    applicationMap.resource = resourceMap
+    json(applicationMap)
+    println json.toPrettyString()
+
+    return null
+  }
+
+  private Void initializeResourceMap(Map resourceMap) {
+    resourceMap.google_compute_instance_group_manager = [:]
+    resourceMap.google_compute_instance_template = [:]
+    resourceMap.google_compute_target_pool = [:]
+    resourceMap.google_compute_forwarding_rule = [:]
+    resourceMap.google_compute_http_health_check = [:]
+    resourceMap.google_compute_autoscaler = [:]
+    resourceMap.google_compute_firewall = [:]
+
+    return null
+  }
+
+  /*
+  Need to do this or terraform complains
+   */
+  private Void cleanUpResourceMap(Map resourceMap) {
+    if (numInstanceGroupManagers == 0) {
+      resourceMap.remove("google_compute_instance_group_manager")
+    }
+    if (numInstanceTemplates == 0) {
+      resourceMap.remove("google_compute_instance_template")
+    }
+    if (numTargetPools == 0) {
+      resourceMap.remove("google_compute_target_pool")
+    }
+    if (numForwardingRules == 0) {
+      resourceMap.remove("google_compute_forwarding_rule")
+    }
+    if (numHealthChecks == 0) {
+      resourceMap.remove("google_compute_http_health_check")
+    }
+    if (numAutoscalers == 0) {
+      resourceMap.remove("google_compute_autoscaler")
+    }
+    if (numFirewalls == 0) {
+      resourceMap.remove("google_compute_firewall")
+    }
+  }
+
+  /*
+   * Adds the server group into a resource map that is in Terraform format. Returns false if the server group is invalid.
+   */
+  private Boolean addServerGroupToResourceMap(GoogleServerGroup.View serverGroup, Map resourceMap) {
+
+    def serverGroupMap = [:]
+    if (serverGroup.name) {
+      serverGroupMap.name = serverGroup.name
+      serverGroupMap.base_instance_name = serverGroup.name
+    } else {
+      //required
+      return false
```

Comment:
If this is required (and will fail without it), it's probably better to throw an exception with a descriptive reason here. It will be surfaced by Gate when the operation fails.